lib/widgets/downloadable_resources_page.dart

```diff
@@ -0,0 +1,293 @@
+//
+// Copyright 2025 Esri
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import 'dart:async';
+import 'dart:io';
+import 'package:arcgis_maps/arcgis_maps.dart';
+import 'package:arcgis_maps_sdk_flutter_samples/common/download_util.dart';
+import 'package:arcgis_maps_sdk_flutter_samples/models/downloadable_resource.dart';
+import 'package:async/async.dart';
+import 'package:flutter/material.dart';
+import 'package:go_router/go_router.dart';
+import 'package:path/path.dart' as path;
+import 'package:path_provider/path_provider.dart';
+
+typedef OnComplete = void Function(List<String>);
+
+/// A page that handles downloading resources required by a sample before opening it.
+///
+/// This page follows the workflow:
+/// - Display sample name and downloadable resources requirement
+/// - Allow user to start/cancel download
+/// - Show progress during download
+/// - Navigate to sample once download is complete
+class DownloadableResourcesPage extends StatefulWidget {
+  const DownloadableResourcesPage({
+    required this.sampleTitle,
+    required this.resources,
+    required this.onComplete,
+    super.key,
+  });
+
+  final String sampleTitle;
+  final List<DownloadableResource> resources;
+  final OnComplete onComplete;
+
+  @override
+  State<DownloadableResourcesPage> createState() =>
+      _DownloadableResourcesPageState();
+}
+
+class _DownloadableResourcesPageState extends State<DownloadableResourcesPage> {
+  var _isDownloading = false;
+  var _isComplete = false;
+  var _isCancelled = false;
+  var _progress = 0.0;
+  
+  // To manage cancellation of download
+  CancelableOperation<List<ResponseInfo>>? _cancelableDownload;
+
+  @override
+  void initState() {
+    super.initState();
+    _checkIfResourcesExist();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text(widget.sampleTitle),
+        leading: IconButton(
+          icon: const Icon(Icons.arrow_back),
+          onPressed: () {
+            if (_isDownloading) {
+              unawaited(_cancelDownload());
+            }
+            context.pop();
+          },
+        ),
+      ),
+      body: Padding(
```

Comment:
When the data is already downloaded and you're re-opening the sample again from the menu, the downloads content is briefly shown before the sample is loaded, like a flash. I'm wondering if the check whether it's downloaded shouldn't be at a higher level? Rather than displaying the page, checking, and then opening the sample.

@martinoyovo could you maybe help test and weigh in on this?

Reply:
Or does this related to issue 4019?